### DISCUSSION TITLE
How to get available WiFis and connect to them ?

### DISCUSSION THREAD

**Discussion by MultiXaNos** [AvaloniaUI/Avalonia#15184] [Q&A]
Hi everybody,

I'm currently working on an app which is running on Linux DRM.
I'm wondering how I can get available list of WiFis and after that how to programmaticaly connect to one of them ?
Does somebody already does something likes this ? There is no resources on this topic over internet...

Thans in advance for help !

**Comment by thevortexcloud** [comment]
There are many networking tools in Linux. Some are much more complicated than others. Your easiest way is to use some of the CLI tools and just parse the outputs. `nmcli` will probably work if your distro supports network manager. Some (like Ubuntu Server) instead use Netplan and systemd networking. Network manager though is considered to be fairly old. `iw` is a newer tool people can use. But it requires root. But  you can for example do: ` sudo iw dev {NICNAME} scan | grep SSID` to see a list of networks.

 You can also parse some of the sysfs data under `/sys/class/net/`.  

How you actually connect ranges from simple to very complicated. It really depends on how your distro is configured. EG you can use config files with wpa_supplicant or netplan.

https://wiki.archlinux.org/title/Wpa_supplicant

I personally normally just use `nmtui` due to how simple it is but as the name implies, it's a terminal UI not a scripting interface.

The Arch wiki covers a lot of your options.

https://wiki.archlinux.org/title/Network_configuration/Wireless

But theoretically you can just use `nmcli` for most of this.

**Comment by wsficke** [comment]
The usual answer is to target a platform API.  This usually involves pulling in .NET platform invocation bindings from some library.

Check this out:

https://github.com/CarteKiwi/Wifi.Linux

If you can't find a .NET library that expresses the bindings you need, introspect the target platform ABI and roll your own with PInvoke: https://developers.redhat.com/blog/2019/03/25/using-net-pinvoke-for-linux-system-functions
https://learn.microsoft.com/en-us/dotnet/standard/native-interop/pinvoke
